platforms/core-execution/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DefaultFileWatcherProbeRegistry.java

```diff
@@ -16,97 +16,172 @@
+        hierarchies
+            .map(File::toPath)
+            .map(this::getProbeForHierarchy)
+            .filter(Objects::nonNull)
+            .filter(x -> x.state == WatchProbeImpl.State.UNARMED)
+            .distinct()
+            .forEach(probe -> {
+                beforeProbeArmed.accept(probe.getLocation(), probe.isSubPathOfWatchableHierarchy());
                 probe.arm();
-            } catch (IOException e) {
-                LOGGER.debug("Could not arm watch probe for hierarchy {}", watchableHierarchy, e);
-            }
-        } else {
-            LOGGER.debug("Did not find watchable hierarchy to arm probe for: {}", watchableHierarchy);
-        }
+            });
     }
 
     @Override
-    public void disarmWatchProbe(File watchableHierarchy) {
-        WatchProbe probe = watchProbesByHierarchy.get(watchableHierarchy);
-        if (probe != null) {
-            probe.disarm();
-        } else {
-            LOGGER.debug("Did not find watchable hierarchy to disarm probe for: {}", watchableHierarchy);
-        }
+    public Stream<File> unprovenHierarchies() {
+        return Streams.concat(
+                probesByFS.values().stream(),
+                watchProbesByHierarchy.values().stream()
+            )
+            .filter(WatchProbeImpl::leftArmed)
+            .flatMap(watchProbe -> watchProbe.getWatchableHierarchies(unwatchedHierarchies));
     }
 
     /**
      * Triggers a watch probe at the given location if one exists.
      */
     @Override
-    public void triggerWatchProbe(String path) {
-        WatchProbe probe = watchProbesByPath.get(path);
+    public void triggerWatchProbe(Path path) {
+        WatchProbeImpl probe = getProbeForHierarchy(path);
         if (probe != null) {
-            LOGGER.debug("Triggering watch probe for {}", probe.getWatchableHierarchy());
+            LOGGER.debug("Triggering watch probe for {}", probe.getWatchableHierarchies(unwatchedHierarchies));
             probe.trigger();
         }
     }
 
-    @Override
-    public File getProbeDirectory(File hierarchy) {
-        WatchProbe watchProbe = watchProbesByHierarchy.get(hierarchy);
+
+    @Nullable
+    private WatchProbeImpl getProbeForHierarchy(Path watchableHierarchy) {
+        FileStore fileStore = getFileStore(watchableHierarchy);
+        WatchProbeImpl watchProbe;
+        if (fileStore != null) {
+            watchProbe = probesByFS.get(fileStore);
+        } else {
+            watchProbe = watchProbesByHierarchy.get(watchableHierarchy);
+        }
         if (watchProbe == null) {
-            throw new IllegalStateException("Cannot find probe for hierarchy: " + hierarchy);
+            LOGGER.debug("Did not find watchable hierarchy probe for: {}", watchableHierarchy);
         }
-        return watchProbe.getProbeFile().getParentFile();
+        return watchProbe;
     }
 
-    private static class WatchProbe {
+    @Nullable
+    private static FileStore getFileStore(Path file) {
```

Comment:
:x: Let's use `Optional<>`.

Reply:
Why? It's always checked against null...